modules/e2eping/e2eping.js

```diff
@@ -0,0 +1,333 @@
+const logger = getLogger(__filename);
+
+/**
+ * The 'type' of a message which designates an e2e ping request.
+ * @type {string}
+ */
+const E2E_PING_REQUEST = 'e2e-ping-request';
+
+/**
+ * The 'type' of a message which designates an e2e ping response.
+ * @type {string}
+ */
+const E2E_PING_RESPONSE = 'e2e-ping-response';
+
+/**
+ * Saves e2e ping related state for a single JitsiParticipant.
+ */
+class ParticipantState {
+    /**
+     * Creates a ParticipantState
+     * @param {JitsiParticipant} participant
+     * @param {E2ePing} e2eping
+     */
+    constructor(participant, e2eping) {
+        // The JitsiParticipant
+        this.participant = participant;
+
+        // The E2ePing
+        this.e2eping = e2eping;
+
+        // Caches the ID
+        this.id = participant.getId();
+
+        // Recently sent requests
+        this.requests = {};
+
+        // The ID of the last sent request. We just increment it for each new
+        // request. Start at 1 so we can consider only thruthy values valid.
+        this.lastRequestId = 1;
+
+        // If the data channel was already open (this is likely a participant
+        // joining an existing conference) send a request immediately.
+        if (e2eping.dataChannelOpened) {
+            this.sendRequest();
+        }
+
+        this.pingInterval = window.setInterval(
+            this.sendRequest, e2eping.pingIntervalMs);
+        this.analyticsInterval = window.setTimeout(
+            this.maybeSendAnalytics, this.e2eping.analyticsIntervalMs);
+    }
+
+    /**
+     * Clears the interval which sends pings.
+     * @type {*}
+     */
+    clearIntervals = function() {
+        if (this.pingInterval) {
+            window.clearInterval(this.pingInterval);
+        }
+        if (this.analyticsInterval) {
+            window.clearInterval(this.analyticsInterval);
+        }
+    }.bind(this); // eslint-disable-line no-invalid-this
+
+    /**
+     * Sends the next ping request.
+     * @type {*}
+     */
+    sendRequest = function() {
+        const requestId = this.lastRequestId++;
+        const requestMessage = {
+            type: E2E_PING_REQUEST,
+            id: requestId
+        };
+
+        this.e2eping.sendMessage(requestMessage, this.id);
+        this.requests[requestId] = {
+            id: requestId,
+            timeSent: window.performance.now()
+        };
+
+    }.bind(this); // eslint-disable-line no-invalid-this
+
+    /**
+     * Handles a response from this participant.
+     * @type {*}
+     */
+    handleResponse = function(response) {
+        const request = this.requests[response.id];
+
+        if (request) {
+            request.rtt = window.performance.now() - request.timeSent;
+            this.e2eping.conference.eventEmitter.emit(
+                E2ePingEvents.E2E_RTT_CHANGED,
+                this.participant,
+                request.rtt);
+        }
+
+        this.maybeSendAnalytics();
+    }.bind(this); // eslint-disable-line no-invalid-this
```

Comment:
why define the functions this way instead of normal function definitions?

Reply:
I initially tried this:
```
handleRespone(response) {
// ...
}
```

It resulted in them not being properly bound :shrug: